proposals/ref-struct-interfaces.md

```diff
@@ -277,6 +285,418 @@ Specifically by using the `CorGenericParamAttr.gpAllowByRefLike(0x0020)` or `Sys
+```csharp
+ref struct S2 : IAsyncDisposable
+{
+    ValueTask IAsyncDisposable.DisposeAsync()
+    {
+    }
+}
+
+class C
+{
+    static async Task Main()
+    {
+        await using (new S2())
+        {
+        } // S2.IAsyncDisposable.DisposeAsync
+    }
+}
+```
+
+Note that preference is given to a `DisposeAsync` method that implements the pattern, and only if one is not found, `IAsyncDisposable`
+implementation is used.
+
+A pattern `DisposeAsync` method will be recognized on a type parameter that `allows ref struct` as it is recognized on
+type parameters without that constraint today.
+
+```csharp
+interface IMyAsyncDisposable
+{
+    ValueTask DisposeAsync();
+}
+
+class C
+{
+    static async Task Test<T>() where T : IMyAsyncDisposable, new(), allows ref struct
+    {
+        await using (new T())
+        {
+        } // IMyAsyncDisposable.DisposeAsync
+    }
+}
+```
+
+A `using` statement will recognize and use implementation of `IAsyncDisposable` interface when resource is a type parameter that 
+`allows ref strict`, there is no `DisposeAsync` pattern match, and `IAsyncDisposable` is in type parameter's effective interfaces set.
+```csharp
+interface IMyAsyncDisposable1
+{
+    ValueTask DisposeAsync();
+}
+
+interface IMyAsyncDisposable2
+{
+    ValueTask DisposeAsync();
+}
+
+class C
+{
+    static async Task Test<T>() where T : IMyAsyncDisposable1, IMyAsyncDisposable2, IAsyncDisposable, new(), allows ref struct
+    {
+        await using (new T())
+        {
+            System.Console.Write(123);
+        } // IAsyncDisposable.DisposeAsync
+    }
+}
+```
+
+### `foreach` statement
+
+A `foreach` statement will recognize and use implementation of ```IEnumerable<T>```/```IEnumerable``` interface when collection is a ref struct.
+```csharp
+ref struct S : IEnumerable<int>
+{
+    IEnumerator<int> IEnumerable<int>.GetEnumerator() {...}
+    System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() {...}
+}
+
+class C
+{
+    static void Main()
+    {
+        foreach (var i in new S()) // IEnumerable<int>.GetEnumerator
+        {
+        }
+    }
+}
+```
+
+A pattern `GetEnumerator` method will be recognized on a type parameter that `allows ref struct` as it is recognized on
+type parameters without that constraint today.
+
+```csharp
+interface IMyEnumerable<T>
+{
+    IEnumerator<T> GetEnumerator();
+}
+
+class C
+{
+    static void Test<T>(T t) where T : IMyEnumerable<int>, allows ref struct
+    {
+        foreach (var i in t) // IMyEnumerable<int>.GetEnumerator
+        {
+        }
+    }
+}
+```
+
+A `foreach` statement will recognize and use implementation of ```IEnumerable<T>```/```IEnumerable``` interface when collection is a type parameter that 
+`allows ref strict`, there is no `GetEnumerator` pattern match, and ```IEnumerable<T>```/```IEnumerable``` is in type parameter's effective interfaces set.
```

Comment:
```suggestion
`allows ref struct`, there is no `GetEnumerator` pattern match, and ```IEnumerable<T>```/```IEnumerable``` is in type parameter's effective interfaces set.
```

Reply:
> A `foreach` statement will recognize and use implementation of ```IEnumerable<T>```/```IEnumerable``` interface when collection is a type parameter that 
> `allows ref strict`, there is no `GetEnumerator` pattern match, and ```IEnumerable<T>```/```IEnumerable``` is in type parameter's effective interfaces set.

Currently, it looks like we use `IEnumerable<T>` even if there is a `GetEnumerator` pattern (available through an interface) - see [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgDABFAjAFgNwFgBQWCWA7AFwFMAnAMwEMBjIuASQFkBPAUTwFcBbUigIwBsiiADwAVAHxYA3ljhz6bLjwIB7EsPwFxcAOJECi7iQqqSACgCUGTAF8sWKACY4AYWmz5SAGwJkcUUQAzgRi4maicAQWcADuABaktBEg9MyGPAJCGoTiADQIiADM2VoecjKY8lVwZGpE1HFwZgBuFCRwOB14kdEA9L0FxZriAHR6BhxGJmpl1RXV8naVi1g2QA).

Reply:
> > A `foreach` statement will recognize and use implementation of `IEnumerable<T>`/`IEnumerable` interface when collection is a type parameter that
> > `allows ref strict`, there is no `GetEnumerator` pattern match, and `IEnumerable<T>`/`IEnumerable` is in type parameter's effective interfaces set.
> 
> Currently, it looks like we use `IEnumerable<T>` even if there is a `GetEnumerator` pattern (available through an interface) - see [sharplab.io](https://sharplab.io/#v2:CYLg1APgAgDABFAjAFgNwFgBQWCWA7AFwFMAnAMwEMBjIuASQFkBPAUTwFcBbUigIwBsiiADwAVAHxYA3ljhz6bLjwIB7EsPwFxcAOJECi7iQqqSACgCUGTAF8sWKACY4AYWmz5SAGwJkcUUQAzgRi4maicAQWcADuABaktBEg9MyGPAJCGoTiADQIiADM2VoecjKY8lVwZGpE1HFwZgBuFCRwOB14kdEA9L0FxZriAHR6BhxGJmpl1RXV8naVi1g2QA).

I can see how this could be confusing, but there is no pattern match in this case due to an ambiguity.